boot/detect.go

```diff
@@ -49,21 +53,40 @@ func (d Detect) Detect(context libcnb.DetectContext) (libcnb.DetectResult, error
 				},
 				Requires: []libcnb.BuildPlanRequire{
 					{Name: PlanEntryJVMApplication},
-					{Name: PlanEntrySpringBoot},
-				},
+					{Name: PlanEntrySpringBoot}				},
 			},
 		},
 	}
-	manifest, err := libjvm.NewManifest(context.Application.Path)
-	if err != nil {
-		return libcnb.DetectResult{}, fmt.Errorf("unable to read manifest in %s\n%w", context.Application.Path, err)
-	}
 
 	cr, err := libpak.NewConfigurationResolver(context.Buildpack, nil)
 	if err != nil {
 		return libcnb.DetectResult{}, fmt.Errorf("unable to create configuration resolver\n%w", err)
 	}
 
+	if sherpa.ResolveBool("BP_JVM_CDS_ENABLED") {
+
+		result = libcnb.DetectResult{
+			Pass: true,
+			Plans: []libcnb.BuildPlan{
+				{
+					Provides: []libcnb.BuildPlanProvide{
+						{Name: PlanEntrySpringBoot},
+					},
+					Requires: []libcnb.BuildPlanRequire{
+						{Name: PlanEntryJVMApplication},
+						{Name: PlanEntrySpringBoot},
+						// Require a JRE at build time to perform CDS training run
+						{Name: PlanEntryJRE, Metadata: map[string]interface{}{"build": true}},
```

Comment:
Right, but you should have a JDK, at least if you're building from source code. I suspect if you're building from pre-compiled assets then you wouldn't have one, but I can't recall off the top of my head. That's why I was asking about that case specifically.

The reason I ask is because it seems like any time we make changes to the build plan, it's a whole thing. There's a very high degree of probability that it messes with something in unintended ways because the build plan is complicated, covering many different scenarios, and affecting all of the buildpacks not just this one (and not necessarily just Paketo buildpacks).

I don't know what the expectations or needs are here but I would offer a suggestion, that the buildpacks only attempt to generate the CDS training data if they are building from source code. 

If you were to do it this way. That would make things less complicated overall, it prevents you from needing to change the build plan, and I think that you could make a case that if someone is already building externally, which is more work for them, then they want more control and would probably also want more control over how the CDS training data is generated. In fact, if we don't support passing it in, I'm pretty confident that at some point we're going to get a request to allow this.

My suggestion in short, if you want the Easy Button then you build from source and build packs do it all. If you want to control how things are done, then you build and generate a CDS file and providing us the artifacts and we just make a container out of them.

Building with the Spring Boot build tools is something I don't know about though. That's definitely a use case to consider. I'm kind of assuming that there's a Maven/Gradle step that can do this, so when using that tool it could pass both the JAR and the CDS training file to buildpacks. If that's not the case, then it probably merits a discussion at least. Is that easy/hard to do? Do we think users might want to pre-generate this file and hand it off to buildpacks? etc...

Reply:
> that the buildpacks only attempt to generate the CDS training data if they are building from source code.
well, Spring Boot Gradle and Maven plugins do not do that; and I believe they're a pretty common way to consume the `spring-boot` buildpack; so I don't think we want to limit ourselves to source code only.

> then they want more control and would probably also want more control over how the CDS training data is generated

I don't agree with that: we expect most CDS users to be your average Spring Boot user, going to start.spring.io and then setting an env. variable (maybe it's even gonna be provided for them eventually) and obseve 1.5x faster startup times.
We don't believe users will want to tweak CDS; just don't use the Spring Boot buildpack if you're an expert?

> I'm kind of assuming that there's a Maven/Gradle step that can do this
Pretty sure there's not. I remember we asked them (@mhalbritter and @snicoll ) - not exactly that question, but on the topic of "please output something else than the spring boot jar or plain jar" and since it's an opinionated framework, they prefer limiting the possible outputs to the strict necessary.

Reply:
>Pretty sure there's not. I remember we asked them (@mhalbritter and @snicoll ) - not exactly that question, but on the topic of "please output something else than the spring boot jar or plain jar" and since it's an opinionated framework, they prefer limiting the possible outputs to the strict necessary.

How would you do CDS without buildpacks then? Is the user left to sort out the Java commands? Maybe there isn't, but I've come to expect handy boot commands like `./gradlew bootRun` that just do things like this (and how native image builds just work). I was thinking something might exist for CDS too.

Reply:
That's correct, there's no support on the Maven or Gradle plugins for this. You can [read more in the documentation](https://docs.spring.io/spring-boot/reference/deployment/efficient.html#deployment.efficient.cds).

Reply:
While I am definitely familiar with the woes of touching the build-plan as Dan mentions, I made this change (when I realised we had no `java` available to run the training-run command), somewhat safe (?) in the knowledge that:

* it is behind the check for opting-in to CDS via the variable - which doesn't remove the risk of issues but narrows the scope
* tl;dr - I believe overall we are expanding plan requirement rather than mixing across current states, the latter being more prone to problems I think. 
  * As far as I could tell nothing else currently requires a JRE at build-time - tomcat/liberty etc require it at run time for obvious reasons. This results in this change **adding** to the Metadata for the JRE-at-runtime entry that buildpacks such as Executable Jar request by default: `Buildpack Plan: {Entries:[{Name:jre Metadata:map[launch:true]} {Name:jre Metadata:map[build:true]}]}`. If building from source, the same appears alongside the build-system buildpack's requested JDK `Name:jdk Metadata:map[]`

Reply:
Sounds good @pivotal-david-osullivan. I agree that this seems lower risk than some of the changes in the past, if for no other reason than that users have to opt-in before they would be impacted.

> As far as I could tell nothing else currently requires a JRE at build-time - tomcat/liberty etc require it at run time for obvious reasons.

+1 I can't think of anything that does either.